assets/src/loader.rs

```diff
@@ -0,0 +1,173 @@
+    /// given path.
+    pub fn new(alloc: &Allocator, directory: PathBuf, pool: Arc<ThreadPool>) -> Self {
+        Loader {
+            contexts: Default::default(),
+            directory: Arc::new(Directory::new(alloc, directory)),
+            pool: pool,
+        }
+    }
+
+    /// Registers an asset and inserts a context into the map.
+    pub fn register<A>(&mut self, context: A::Context)
+        where A: Asset + 'static,
+              A::Context: 'static
+    {
+        self.contexts
+            .insert(TypeId::of::<A>(), Box::new(Arc::new(context)));
+    }
+
+    /// Loads an asset with a given format from the default (directory) store.
+    /// If you want to load from a custom source instead, use `load_from`.
+    ///
+    /// The actual work is done on a worker thread, thus this method immediately returns
+    /// a future.
+    pub fn load<A, F, I>(&self,
+                         id: I,
+                         format: F)
+                         -> AssetFuture<A, LoadError<A::Error, F::Error, IoError>>
+        where A: Asset + Send + 'static,
+              A::Context: Send + Sync + 'static,
+              A::Error: Send + 'static,
+              F: Format<Data = A::Data> + Send + 'static,
+              F::Error: Send + 'static,
+              I: AsRef<str> + Send + 'static
+    {
+        self.load_from::<A, F, _, _>(id, format, self.directory.clone())
+    }
+
+    /// Loads an asset with a given id and format from a custom store.
+    /// The actual work is done on a worker thread, thus this method immediately returns
+    /// a future.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the asset wasn't registered.
+    pub fn load_from<A, F, I, S>(&self,
+                                 id: I,
+                                 format: F,
+                                 store: Arc<S>)
+                                 -> AssetFuture<A, LoadError<A::Error, F::Error, S::Error>>
+        where A: Asset + Send + 'static,
+              A::Context: Send + Sync + 'static,
+              A::Error: Send + 'static,
+              F: Format<Data = A::Data> + Send + 'static,
+              F::Error: Send + 'static,
+              I: AsRef<str> + Send + 'static,
+              S: Store + Send + Sync + 'static,
+              S::Error: Send
+    {
+        let context = self.contexts
+            .get(&TypeId::of::<A>())
+            .expect("Assets need to be registered");
+
+        let context: &Arc<A::Context> = context.downcast_ref().unwrap();
+
+        load_asset_future::<A, F, I, S>(context.clone(), format, id, store, &*self.pool)
+    }
+}
+
+/// Loads an asset with a given context, format, specifier and storage right now.
+pub fn load_asset<A, F, S>(context: &A::Context,
+                           format: &F,
+                           name: &str,
+                           storage: &S)
+                           -> Result<A, LoadError<A::Error, F::Error, S::Error>>
```

Comment:
Given that typedefs don't support bounds AFAIK, this wouldn't be much shorter: 
`AssetResult<A, A::Error, F::Error, S::Error>`. And it hides the real type, so probably more to lookup?